From 3a6a277691e4618f33165ee3c0d48eed5b5a2490 Mon Sep 17 00:00:00 2001
From: Karthik Kambatla <kasha@cloudera.com>
Date: Thu, 30 May 2013 11:43:58 -0700
Subject: [PATCH 072/596] MR1: CLOUDERA-BUILD. Job History search for
 non-existent dates results in error

Reason: Customer request/ Bug fix
(cherry picked from commit 1c1b318a626bed66f610c9db4882d377dc3aa9ca)
(cherry picked from commit 43f4c9f6a2b817d3f59206b3197cc8bcd94e1125)
(cherry picked from commit 44066ba9b13ac009e87252136b46b264142df9ba)
(cherry picked from commit 8f5c5ed8f7e87044cf6ebc6c887c24e056cc70a3)
---
 .../org/apache/hadoop/mapred/JobHistory.java       |   28 +++++++++++++++-----
 1 file changed, 22 insertions(+), 6 deletions(-)

diff --git a/hadoop-mapreduce1-project/src/mapred/org/apache/hadoop/mapred/JobHistory.java b/hadoop-mapreduce1-project/src/mapred/org/apache/hadoop/mapred/JobHistory.java
index 76cf446..426b401 100644
--- a/hadoop-mapreduce1-project/src/mapred/org/apache/hadoop/mapred/JobHistory.java
+++ b/hadoop-mapreduce1-project/src/mapred/org/apache/hadoop/mapred/JobHistory.java
@@ -20,6 +20,7 @@
 
 import java.io.BufferedReader;
 import java.io.File;
+import java.io.FileNotFoundException;
 import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStreamReader;
@@ -899,7 +900,22 @@ public synchronized void handle(Map<Keys, String> values){
       throws IOException {
     return localGlobber(fs, root, tail, filter, null);
   }
-  
+
+  private static FileStatus[] nullToEmpty(FileStatus[] result) {
+    return result == null ? new FileStatus[0] : result;
+  }
+
+  private static FileStatus[] listFilteredStatus
+      (FileSystem fs, Path root, PathFilter filter)
+      throws IOException {
+    try {
+      return filter == null ? fs.listStatus(root) : fs.listStatus(root,
+          filter);
+    } catch (FileNotFoundException fnfe) {
+      // file doesn't exist
+      return null;
+    }
+  }
 
   // hasMismatches is just used to return a second value if you want
   // one.  I would have used MutableBoxedBoolean if such had been provided.
@@ -907,11 +923,12 @@ public synchronized void handle(Map<Keys, String> values){
     (FileSystem fs, Path root, String tail, PathFilter filter, AtomicBoolean hasFlatFiles)
       throws IOException {
     if (tail.equals("")) {
-      return filter == null ? fs.listStatus(root) : fs.listStatus(root, filter);
+      return nullToEmpty(listFilteredStatus(fs, root, filter));
     }
 
       if (tail.startsWith("/*")) {
-        Path[] subdirs = filteredStat2Paths(fs.listStatus(root), true, hasFlatFiles);
+        Path[] subdirs = filteredStat2Paths(nullToEmpty(listFilteredStatus
+            (fs, root, null)), true, hasFlatFiles);
 
         FileStatus[][] subsubdirs = new FileStatus[subdirs.length][];
 
@@ -944,9 +961,8 @@ public synchronized void handle(Map<Keys, String> values){
         int split = tail.indexOf('/', 1);
 
         if (split < 0) {
-          return (filter == null
-                  ? fs.listStatus(new Path(root, tail.substring(1)))
-                  : fs.listStatus(new Path(root, tail.substring(1)), filter));
+          return nullToEmpty
+              (listFilteredStatus(fs, new Path(root, tail.substring(1)), filter));
         } else {
           String thisSegment = tail.substring(1, split);
           String newTail = tail.substring(split);
-- 
1.7.9.5

